File: kanvas/src/commonMain/kotlin/com/cws/kanvas/core/async/ThreadPool.kt
package com.cws.kanvas.core.async

class ThreadPool(
    name: String,
    priority: Int,
    size: Int,
    private var start: Boolean = false,
) {

    private var running = false
    private val tasks = ConcurrentQueue<Task>(100)
    private val threads: Array<PlatformThread> = Array(size) {
        PlatformThread(
            name = "$name-1",
            priority = priority,
            task = ::runLoop,
            start = start,
        )
    }

    fun start() {
        if (!start) {
            start = true
            threads.forEach { thread ->
                thread.start()
            }
        }
    }

    fun release() {
        start = false
        running = false
    }

    fun submit(task: Task) {
        tasks.push(task)
    }

    private fun runLoop() {
        running = true
        while (running) {
            val task = tasks.pop()
            task?.invoke()
        }
    }

}

File: kanvas-vk/src/cpp/vk/core/thread_pool.cpp
//
// Created by cheerwizard on 06.07.25.
//

#include "thread_pool.hpp"

ThreadPool::ThreadPool(u32 size) {
    for (int i = 0; i < size; i++) {
        std::thread thread([&] { runLoop(); });
        thread.detach();
    }
}

ThreadPool::~ThreadPool() {
    running = false;
}

void ThreadPool::submit(const Task &task) {
    taskQueue.push(task);
}

void ThreadPool::runLoop() {
    running = true;
    while (running) {
        Task task;
        taskQueue.pop(task);
        task.task();
    }
}